File: src/main/java/com/interview/binarysearch/BinarySearch.java
package com.interview.binarysearch;

/**
 * Regular binary search
 */
public class BinarySearch {

    public int search(int input[],int search){
        int low = 0;
        int high = input.length-1;
        int mid = 0;
        while(low <= high){
            mid = (low + high)/2;
            if(input[mid] == search){
                return mid;
            }else if(input[mid] < search){
                low = mid+1;
            }else{
                high = mid-1;
            }
        }
        return -1;
    }
    
    public static void main(String args[]){
        BinarySearch bs = new BinarySearch();
        int arr1[] = {1,2,4,5,7,8};
        System.out.println(bs.search(arr1, -1));
        System.out.println(bs.search(arr1, 1));
        System.out.println(bs.search(arr1, 8));
        System.out.println(bs.search(arr1, 2));
    }
}

File: src/com/interview/binarysearch/BinarySearch.py
def search(input_list, num): 
  #return index if found, -1 if not 
  #iterative to avoid space complexity implications of the recursive solution 
  low = 0 
  high = len(input_list)
  middle = (low + high)/2 
  
  while low < high: 
    if input_list[middle] == num: 
      return middle 
    elif input_list[middle] < num: 
      low = middle
      middle = (low + high)/2 
    else: 
       high = middle
       middle = (low + high)/2 
  return -1 
       
def main(): 
  input_list = [1, 2, 4, 5, 7, 8]
  print search(input_list, -1)
  print search(input_list, 1)
  print search(input_list, 8)
  print search(input_list, 2)
